### PR TITLE
Implement Staking Formula for Polkapool

### PR DESCRIPTION
**Pull request by Gauthamastro**
This PR implements the staking formula and clears some lint errors

### CODE REVIEW

File: pallets/polkapool/src/lib.rs [hunk 1]
```diff
@@ -19,131 +19,160 @@
...
+        CurrencyId=AssetId,
+        Balance=Self::Balance,
+    > + MultiLockableCurrency<Self::AccountId>;
     /// Min amount that must be staked
     type MinStakeAmount: Get<Self::Balance>;
     /// Maximum allowed Feeless Transactions in a block
     type MaxAllowedWeight: Get<Weight>;
     /// Min Stake Period
-    type MinStakePeriodPerWeight: Get<u32>;
+    type MinStakePeriod: Get<BlockNumber>;
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:87]: Defines the min amount of time the user needs to stake their tokens, currently set to 28 days

File: pallets/polkapool/src/lib.rs [hunk 2]
```diff
@@ -19,131 +19,160 @@
...
+    /// Max number of stakes per account
+    type MaxStakes: Get<usize>;
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:89]: Max number of feeless transactions that a single account can have at a given time.

File: pallets/polkapool/src/lib.rs [hunk 3]
```diff
@@ -19,131 +19,160 @@
...
     }
 }
 
-#[derive(Decode, Encode, Copy, Clone)]
-pub struct MovingAverage<T: Config> {
-    pub amount: <T as Config>::Balance,
-    pub count: <T as Config>::Balance,
+#[derive(Decode, Encode, Clone)]
+pub struct StakeInfo<T: Config + frame_system::Config> {
+    stakes: VecDeque<Stake<T>>,
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:126]: Defined a vecdeque for storing the different locks aka stakes for different extrinsics

File: pallets/polkapool/src/lib.rs [hunk 4]
```diff
@@ -254,54 +280,43 @@ decl_module! {
...
+            let mut staked_info: StakeInfo<T> = Self::staked_users(who.clone());
+            staked_info.push(stake_amount,stake_period+current_block)?;
 
-            //Todo change to new Formula
-            staked_amount.unlocking_block =  staked_amount.unlocking_block
-            .checked_add(&(call_weight as u32).saturated_into::<T::BlockNumber>().checked_mul(&T::MinStakePeriodPerWeight::get().saturated_into::<T::BlockNumber>()).ok_or(<Error<T>>::Overflow)?)
-            .ok_or(<Error<T>>::Overflow)?;
+             //Reserve stake amount
+            ensure!(T::Currency::reserve(AssetId::POLKADEX, &who, stake_amount).is_ok(),Error::<T>::FailedToMoveBalanceToReserve);
+            // TODO: Replace reserver with T::Currency::set_lock() since that makes more sense
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:308]: Reserve will be replaced with `lock()` for locking the assets

File: pallets/polkapool/src/lib.rs [hunk 5]
```diff
@@ -313,11 +328,14 @@ decl_module! {
...
             ensure!(origin.clone().into().is_ok(),Error::<T>::BadOrigin);
             ensure!(<StakedUsers<T>>::contains_key(&who),Error::<T>::StakeNotFound);
-            let stake = <StakedUsers<T>>::get(&who);
+            let mut stake_info: StakeInfo<T> = <StakedUsers<T>>::get(&who);
             let current_block_no: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
-            ensure!(stake.unlocking_block <= current_block_no,Error::<T>::UnlockingFailedCurrentBlockNumberLow);
-            T::Currency::unreserve(AssetId::POLKADEX, &who, stake.staked_amount);
-            Self::deposit_event(RawEvent::Unstaked(who,stake.staked_amount));
+            let mut total: T::Balance = 0u128.saturated_into();
+            for stake in stake_info.claimable_stakes(current_block_no){
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:334]: here, user will be able to claim all the claimable tokens in one go

File: pallets/polkapool/src/lib.rs [hunk 6]
```diff
@@ -327,11 +345,28 @@ decl_module! {
         pub fn slash_stake(origin, account: T::AccountId) -> DispatchResult {
             let origin = <T as Config>::Origin::from(origin);
             T::GovernanceOrigin::ensure_origin(origin)?;
-            let stake = <StakedUsers<T>>::get(&account);
-            <StakedUsers<T>>::remove(&account);
-            T::Currency::slash(AssetId::POLKADEX,&account,stake.staked_amount);
-            Self::deposit_event(RawEvent::StakeSlashed(account,stake.staked_amount));
+            let mut stake_info: StakeInfo<T> = <StakedUsers<T>>::take(&account);
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:348]: The governance can slash all the staked tokens of an account completely if this account is found to be spamming the network.

File: pallets/polkapool/src/lib.rs [hunk 7]
```diff
@@ -327,11 +345,28 @@ decl_module! {
...
             Ok(())
         }
     }
 }
+
+impl<T: Config> Module<T> {
+    // Calculates the stake amount and staking period
+    // The staking period will vary between 28-56 days. The heavier the transaction, longer the period.
+    // A transaction consuming the full allocated weight lock the tokens for 56 days.
+    pub fn calculate_stake_params(stake_price: T::Balance, call_weight: Weight) -> (T::Balance, T::BlockNumber) {
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:360]: Assume the fraction of total allocated block weight used by the call is `x`, then the staking period is set as  `MinStakePeriod*x + MinStakePeriod`. Therefore, if a call consumes the complete weight, then the staking period will be `2*MinStakePeriod`

Comment (Gauthamastro) [reply pallets/polkapool/src/lib.rs:360]: I am still researching new ways to calculate the stake amount and stake period based on the frequency of use, call weight, etc

**Review by zktony — APPROVED** [review]

**Pull request merged**